src/number_theory/pell_general.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2023 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Geißer, Michael Stoll
+-/
+
+import tactic.qify
+import data.zmod.basic
+import number_theory.diophantine_approximation
+
+/-!
+# Pell's Equation
+
+We prove the following
+
+**Theorem.** Let $d$ be a positive integer that is not a square. Then the equation
+$x^2 - d y^2 = 1$ has a nontrivial (i.e., with $y \ne 0$) solution in integers.
+
+See `pell.ex_nontriv_sol`.
+
+This is the beginning of a development that aims at providing all of the essential theory
+of Pell's Equation for general $d$ (as opposed to the contents of `number_theory.pell`,
+which is specific to the case $d = a^2 - 1$ for some $a > 1$).
+
+## References
+
+* [K. Ireland, M. Rosen, *A classical introduction to modern number theory*
+   (Section 17.5)][IrelandRosen1990]
+
+## Tags
+
+Pell's equation
+-/
+
+namespace pell
+
+section existence
+
+open set real
+
+/-- If `d` is a positive integer that is not a square, then there is a nontrivial solution
+to the Pell equation `x^2 - d*y^2 = 1`. -/
+theorem ex_nontriv_sol {d : ℤ} (h₀ : 0 < d) (hd : ¬ is_square d) :
+  ∃ x y : ℤ, x ^ 2 - d * y ^ 2 = 1 ∧ y ≠ 0 :=
+begin
+  let ξ : ℝ := sqrt d,
+  have hξ : irrational ξ,
+  { refine irrational_nrt_of_notint_nrt 2 d (sq_sqrt $ int.cast_nonneg.mpr h₀.le) _ two_pos,
+    rintro ⟨x, hx⟩,
+    refine hd ⟨x, @int.cast_injective ℝ _ _ d (x * x) _⟩,
+    rw [← sq_sqrt $ int.cast_nonneg.mpr h₀.le, int.cast_mul, ← hx, sq], },
+  obtain ⟨M, hM₁⟩ := exists_nat_gt (2 * ξ + 1),
+  have hM : {q : ℚ | |(q.1) ^ 2 - d * (q.2) ^ 2| < M}.infinite,
+  { refine infinite.mono (λ q (h : |ξ - q| < 1 / q.2 ^ 2), _)
+      (infinite_rat_abs_sub_lt_one_div_denom_sq_of_irrational hξ),
+    replace h : |ξ - q.1 / q.2| < 1 / q.2 ^ 2 := by {convert h, exact_mod_cast q.num_div_denom},
+    have hq : 0 < (q.2 : ℝ) := nat.cast_pos.mpr q.pos,
+    have H : (d : ℝ) * q.2 ^ 2 - q.1 ^ 2 =
+               (ξ - q.1 / q.2) * q.2 ^ 2 * (2 * ξ - (ξ - q.1 / q.2)),
+    { field_simp [hq.ne'], rw ← sq_sqrt (int.cast_pos.mpr h₀).le, ring, },
+    have claim : |(q.1 : ℝ) ^ 2 - d * q.2 ^ 2| < M,
+    { rw [abs_sub_comm, H, abs_mul, abs_mul, abs_of_nonneg (sq_nonneg (q.2 : ℝ))],
+      refine (mul_le_mul_of_nonneg_right (mul_le_mul_of_nonneg_right h.le $ sq_nonneg _) $
+        abs_nonneg _).trans_lt _,
+      rw [one_div_mul_cancel (sq_pos_of_pos hq).ne', one_mul],
+      refine (((abs_sub _ _).trans_lt $ add_lt_add_left h _).trans_le $
+        add_le_add_left (_ : _ ≤ 1) _).trans _,
+      { refine (one_div_le zero_lt_one $ sq_pos_of_pos hq).mp _,
+        rw [div_one, one_le_sq_iff hq.le],
+        exact_mod_cast q.pos, },
+      { rwa abs_of_nonneg (mul_nonneg (zero_le_two : (0 : ℝ) ≤ 2) (sqrt_nonneg _)), } },
+    exact_mod_cast claim, },
```

Comment:
Slight golf:
```suggestion
  obtain ⟨M, hM⟩ := exists_int_gt (2 * |ξ| + 1),
  have hM : {q : ℚ | |q.1 ^ 2 - d * q.2 ^ 2| < M}.infinite,
  { refine infinite.mono (λ q h, _) (infinite_rat_abs_sub_lt_one_div_denom_sq_of_irrational hξ),
    have h0 : 0 < (q.2 : ℝ) ^ 2 := pow_pos (nat.cast_pos.mpr q.pos) 2,
    have h1 : (q.num : ℝ) / (q.denom : ℝ) = q := by exact_mod_cast q.num_div_denom,
    rw [mem_set_of, abs_sub_comm, ← @int.cast_lt ℝ, ← div_lt_div_right (abs_pos_of_pos h0)],
    push_cast,
    rw [← abs_div, abs_sq, sub_div, mul_div_cancel _ h0.ne',
        ← div_pow, h1, ← sq_sqrt (int.cast_pos.mpr h₀).le, sq_sub_sq, abs_mul, ← mul_one_div],
    refine mul_lt_mul'' (((abs_add ξ q).trans _).trans_lt hM) h (abs_nonneg _) (abs_nonneg _),
    rw [two_mul, add_assoc, add_le_add_iff_left, ← sub_le_iff_le_add'],
    rw [mem_set_of, abs_sub_comm] at h,
    refine (abs_sub_abs_le_abs_sub (q : ℝ) ξ).trans (h.le.trans _),
    rw [div_le_one h0, one_le_sq_iff_one_le_abs, nat.abs_cast, nat.one_le_cast],
    exact q.pos },
```

Reply:
Thanks! I thought I had golfed it down as far as possible, but there is always room for improvement...